src/main/java/gregtech/api/graphnet/alg/SinglePathAlgorithm.java

```diff
@@ -0,0 +1,89 @@
+                                                                                       NetPathMapper<Path> remapper) {
+        if (recomputeEveryCall) {
+            return (graph, testObject, simulator, queryTick) -> {
+                graph.prepareForAlgorithmRun(testObject, simulator, queryTick);
+                List<GraphEdge> graphEdges = new ObjectArrayList<>();
+                List<GraphVertex> nodes = new ObjectArrayList<>();
+                Results results = compute(source, nodes, graphEdges);
+                if (!results.valid()) return SimpleIteratorFactories.getSingletonIterator(null);
+                if (graphEdges.isEmpty()) return SimpleIteratorFactories.getSingletonIterator(remapper.map(source));
+                return SimpleIteratorFactories.getSingletonIterator(remapper.map(nodes, graphEdges, results.weight()));
+            };
+        } else {
+            List<GraphEdge> graphEdges = new ObjectArrayList<>();
+            List<GraphVertex> nodes = new ObjectArrayList<>();
+            Results results = compute(source, nodes, graphEdges);
+            if (!results.valid()) return SimpleIteratorFactories.emptyFactory();
+            if (graphEdges.isEmpty()) return SimpleIteratorFactories.fromSingleton(remapper.map(source));
+            return SimpleIteratorFactories.fromSingleton(remapper.map(nodes, graphEdges, results.weight()));
+        }
+    }
+
+    private Results compute(GraphVertex source,
+                            List<GraphVertex> nodes, List<GraphEdge> graphEdges) {
+        nodes.add(source);
+        GraphVertex lastNode = null;
+        GraphVertex node = source;
+        GraphEdge graphEdge;
+        double sumWeight = 0;
+        boolean valid = true;
+        while (valid) {
+            Iterator<GraphEdge> i = this.net.getGraph().outgoingEdgesOf(node).iterator();
+            if (!i.hasNext()) break; // we've reached the end, exit the loop while still valid
+            graphEdge = i.next();
+            // if we are directed, we know that the target is the target.
+            // if we aren't directed, we need to see if the graphEdge's target is secretly the source
+            boolean reversedEdge = !this.net.getGraph().isDirected() && graphEdge.getTarget() == node;
+            if ((!reversedEdge && graphEdge.getTarget() == lastNode) ||
+                    (reversedEdge && graphEdge.getSource() == lastNode)) {
+                // current edge points to a previous node, either get the other edge or exit safely.
+                if (i.hasNext()) {
+                    graphEdge = i.next();
+                    reversedEdge = !this.net.getGraph().isDirected() && graphEdge.getTarget() == node;
+                    // we know that the new edge cannot point to the previous node
+                } else break; // we've reached the end, exit the loop while still valid
+            } else if (i.hasNext()) i.next();
```

Comment:
If we reach this else, then only one edge has been burned from the iterator because we stumbled upon a progressing edge immediately. If the node had two edges (which is the expected case for a single path net), then the following call of hasNext() would return true and mark the path invalid. This extra burn is to make sure that only 3+ edges cause invalidation, not 2 edges.